Sources/ObliviousHTTP/BHTTPSerializer.swift

```diff
@@ -88,18 +153,157 @@ public struct BHTTPSerializer {
+    }
 }
 
+// Enum definitions for message, states, and types.
 extension BHTTPSerializer {
-    /// Types of message for binary http serilaisation
+    // Finite State Machine for managing transitions in BHTTPSerializer.
+    public class BHTTPSerializerFSM {
+        private var currentState: BHTTPSerializerState
+
+        init(initialState: BHTTPSerializerState) {
+            self.currentState = initialState
+        }
+
+        func writeRequestHead(
+            _ requestHead: HTTPRequestHead,
+            into buffer: inout ByteBuffer,
+            using serializer: inout BHTTPSerializer
+        ) throws {
+            try self.transition(to: .header)
+            serializer.serializeRequestHead(requestHead, into: &buffer)
+        }
+
+        func writeResponseHead(
+            _ responseHead: HTTPResponseHead,
+            into buffer: inout ByteBuffer,
+            using serializer: inout BHTTPSerializer
+        ) throws {
+            try self.transition(to: .header)
+            serializer.serializeResponseHead(responseHead, into: &buffer)
+        }
+
+        func writeRequestEnd(
+            into buffer: inout ByteBuffer,
+            using serializer: inout BHTTPSerializer
+        ) throws {
+            serializer.endRequest(into: &buffer)
+            try self.transition(to: .end)
+        }
+        func writeBodyChunk(
+            _ body: ByteBuffer,
+            into buffer: inout ByteBuffer,
+            using serializer: inout BHTTPSerializer
+        ) throws {
+            serializer.serializeChunk(body, into: &buffer)
+            try self.transition(to: .chunk)
+        }
+
+        func writeTrailers(
+            _ trailers: HTTPHeaders,
+            into buffer: inout ByteBuffer,
+            using serializer: inout BHTTPSerializer
+        ) throws {
+            serializer.serializeTrailers(trailers, into: &buffer)
+            try self.transition(to: .trailers)
+        }
+
+        private func transition(to state: BHTTPSerializerState) throws {
+            let allowedNextStates: Set<BHTTPSerializerState>
+            switch currentState {
+            case .start:
+                allowedNextStates = [.header]
+            case .header:
+                allowedNextStates = [.chunk, .trailers, .end]
+            case .chunk:
+                allowedNextStates = [.trailers, .end, .chunk]
+            case .trailers:
+                allowedNextStates = [.trailers, .end]
+            case .end:
+                allowedNextStates = []
+            }
+            guard allowedNextStates.contains(state) else {
+                throw ObliviousHTTPError.unexpectedHTTPMessageSection()
+            }
+            currentState = state
+        }
+    }
+
     public enum Message {
-        /// Part of an HTTP request.
         case request(HTTPClientRequestPart)
-        /// Part of an HTTP response.
         case response(HTTPServerResponsePart)
     }
+
+    public struct SerializerType: Equatable {
```

Comment:
Minor nit: it's rarely worth making a type `Equatable`. You should usually jump all the way to `Hashable`: there just aren't many types where equality is well-defined but hashability isn't. While we're here we should also make this type `Sendable`.